### ISSUE TITLE
Copy built-in integration with Express v4 to a standalone npm package like all other integrations; support Express v5 as well

### ISSUE THREAD

**Issue by joergbaier** [apollographql/apollo-server#7928]
`@apollo/server` has express v4 as a dependency. After several years, express is moving forward with releasing v5 (currently published as `next` tag). Having support for v5 here would unblock and encourage people to upgrade.

https://www.npmjs.com/package/express?activeTab=versions
https://github.com/expressjs/express/releases/tag/v5.0.0

At this time, I have not checked for breaking changes in this repo, but wanted to open this issue to get things rolling. Any changes could be published under the `next` tag as well.

**Comment by joergbaier** [comment]
Typing updates for `@types/express` & `@types/express-serve-static-core`
https://github.com/DefinitelyTyped/DefinitelyTyped/pull/70563

**Comment by RobinTail** [comment]
Breaking changes are listed in the [migration guide](https://expressjs.com/en/guide/migrating-5.html), which is not yet up to date. Besides the listed changes there is also Node.js min supported version: 18.
The primary feature is proper handling of asynchronous handlers.

**Comment by glasser** [comment]
I think it would be a good start if community members wanted to publish Express 5 integrations separately (perhaps under the `@as-integrations` npm org / `apollo-server-integrations` GitHub org)!  While we could add a `@apollo/server/express5` directory directly to Apollo Server, it might be challenging to do so in a way that uses different dependencies for two different parts of the same npm module, so getting this in core might be more likely a thing for the next major version of AS, which is not currently scheduled.

**Comment by RobinTail** [comment]
I made types. Enjoy :)

![image](https://github.com/user-attachments/assets/a52c033c-9563-4710-bd68-c9ca6ee55ba2)

**Comment by P4sca1** [comment]
I am using express5 beta with apollo server 4 for the past year without any problems.
After upgrading `@types/express` and `@types/express-serve-static-core` to v5, I needed to override the version in apollo server to avoid type mismatches. e.g. for pnpm:

```json
{
"pnpm": {
                "overrides": {
                        "@types/express": "5.0.0",
                        "@types/express-serve-static-core": "5.0.0"
                }
        }
}
```

**Comment by AndKiel** [comment]
I tried upgrading to express v5 and it looks like apollo-server is incompatible. This check:

https://github.com/apollographql/apollo-server/blob/597f6e1cb7abc65282737ff46d3c008bc8384c25/packages/server/src/express4/index.ts#L47

is invalid because of the following breaking change:

> The req.body property returns undefined when the body has not been parsed. In Express 4, it returns {} by default.

**Comment by glasser** [comment]
I don't think it should be a big surprise that the code in the `express4` directory does not work with Express 5.

The main interface for Apollo Server is the generic interface that works with any web server if you write a short adapter. Apollo Server does ship with one "first party" binding for the most popular and stable web framework, Express v4.  But it's an explicit goal of the AS4 rewrite that it is *not* the job of Apollo Server itself to provide bindings to every web framework out there. That's how things worked in AS3 — the core project needed to have direct integrations to all supported web frameworks and there was no extensibility.  In AS4 we provide a built-in integration with the most popular framework (Express v4) and community packages can provide integrations (typically about a page of code) with other frameworks.

It is likely that at some point in the future, Express v5 may be the most popular and stable web framework, but for today, that's still Express v4. So as mentioned above in https://github.com/apollographql/apollo-server/issues/7928#issuecomment-2354138374 I'd encourage folks on the cutting edge to put together a community package for `@as-integrations/express5`.

Perhaps Apollo Server v5 would incorporate that as its default and shift Express v4 to "community/external". But for now I would assume the place to start is with an integration.

**Comment by cupofjoakim** [comment]
> Apollo Server does ship with one "first party" binding for the most popular and stable web framework, Express v4.
> ... 
> It is likely that at some point in the future, Express v5 may be the most popular and stable web framework, but for today, that's still Express v4.

@glasser Is this the perspective of apollo for node versions too? To only support them once it's the most popular version? Express 4 and 5 are not different frameworks, they are just different versions. 

Does this mean you're closed for contributions of an express 5 adapter too?

**Comment by glasser** [comment]
A library that runs on 0 versions of the runtime is not useful. So we will continue to track versions of the runtime that we support. (Of course, we have other issues like "we are overdue to bumping our major version number and dropping old Node version support" too...)

A library that has zero built-directly-into-the-core-package implementations of the interface it exports is still useful.  Express 4 is an outlier here in that it is very slightly easier to use than all of the other frameworks.

The versions of Express have different API, so building against them require linking in multiple versions of the same library or its DefinitelyTyped libraries. This makes the build more complex and hard to evaluate.

We went through all of this previously with Fastify v2 to v3. It's a lot easier to ship entirely separate adapter packages for different incompatible versions of a framework than to try to make one package that magically works with multiple versions.

The users of Fastify, Hapi, Koa, Next.js, and h3 are fine installing an integration package. It will work fine for Express v5 too.  And now that we know that Express is a moving target, it probably makes sense to put it in on equal footing with all the other web frameworks and make the next major version of `@apollo/server` a low-dependency pure implementation of its interface instead of anointing a single version of a single framework as special.

**Comment by AndKiel** [comment]
When the `req.body` check is removed from `expressMiddleware`, my API seems to work fine, including a route configured after the `ApolloServer` (otherwise, the `req.body` check would prevent it from executing). I don't know why that check is there and how likely is the scenario of someone forgetting to set up the `body-parser`. It looks like the only thing required to make it work with express v5 (?).

**Comment by joergbaier** [comment]
express has tagged v5.1.0 as "latest"
https://github.com/expressjs/express/releases/tag/v5.1.0

**Comment by glasser** [comment]
This is scoped in with the ASv5 work at #7515. I expect that the new Express-v5 package will work with the current ASv4 as well as the forthcoming ASv5.

**Comment by gbresciano** [comment]
thanks @AndKiel this fixed it for me

**Comment by cramatt** [comment]
I was able to workaround this by adding simple middlware: 

```
app.use((req, res, next) => {
    req.body = req.body || {};
    next();
});
```

**Comment by glasser** [comment]
We've now published `@as-integrations/express4` and `@as-integrations/express5` as officially-maintained integration packages, and updated the docs to recommend their use instead of `@apollo/server/express4` (including documenting the Express 5 version as the default one to use for everything other than migrating from old versions of Apollo Server). This puts Express on the same playing field as all other JS web frameworks.

I ended up keeping in the "you forgot to parse the body" warning and just making it a little less sensitive.

`@apollo/server/express4` will be removed in Apollo Server 5 (#7515).
